design-proposals/klip-6-ksql-testing-tool.md

```diff
@@ -0,0 +1,186 @@
+
+```shell
+$ ksql-testing-tool /path/to/passing_test.json
+All tests passed!
+$
+```
+The following is a sample output of the tool when a test fails:
+
+```shell
+$ ksql-testing-tool /path/to/failing_test.json
+Exception in thread "main" java.lang.AssertionError: TestCase name: project-filter - project and filter in file: /path/to/correct_test.json failed while processing output row 0 topic: S1 due to: Expected <1010, 101> with timestamp=0 but was <101, 101> with timestamp=0
+  at io.confluent.ksql.test.commons.TestCase.verifyOutput(TestCase.java:190)
+  at io.confluent.ksql.testingtool.TestRunner.shouldBuildAndExecuteQuery(TestRunner.java:134)
+  at io.confluent.ksql.testingtool.TestRunner.main(TestRunner.java:96)
+Caused by: java.lang.AssertionError: Expected <1010, 101> with timestamp=0 but was <101, 101> with timestamp=0
+  at org.apache.kafka.streams.test.OutputVerifier.compareKeyValueTimestamp(OutputVerifier.java:208)
+  at io.confluent.ksql.test.commons.TestCase.verifyOutput(TestCase.java:177)
+  ... 2 more
+$
+```
+
+As you can see, the error message will include the failure reason.
+
+## Alternative Public API
+
+An alternative to the above option is to separate the test data from the statements. In this case we will have the KSQL statements in a
+ query file similar to the headless execution mode. The rest of the test attributes including input and output data will remain in the test file in JSON format.
+ The following are the query file and the test file in this option:
+
+ Query file:
+ ```SQL
+ CREATE STREAM TEST (ID bigint, NAME varchar, VALUE double) WITH (kafka_topic='test_topic', value_format='DELIMITED', key='ID');
+ CREATE STREAM S1 as SELECT name FROM test where id > 100;
+ ```
+
+
+ Test file:
+
+ ```JSON
+ {
+  "comments": [
+    "Any comments on this test file!"
+  ],
+  "tests": [
+    {
+      "name": "Sample test",
+      "inputs": [
+        {"topic": "test_topic", "key": 0, "value": "0,zero,0.0"},
+        {"topic": "test_topic", "key": 100, "value": "100,100,0.0"},
+        {"topic": "test_topic", "key": 101, "value": "101,101,0.0"}
+      ],
+      "outputs": [
+        {"topic": "S1", "key": 101, "value": "101"}
+      ]
+    }
+  ]
+ }
+ ```
+
+ In this alternative, the testing tool command will receive two named parameters, _--query-file_ and _test-file_.
+ The following is a sample use of the testing tool command in this case:
+
+ ```shell
+ $ ksql-testing-tool --query-file /path/to/query_file.sql --test-file /path/to/test_file.json
+ All tests passed!
+ $
+ ```
+
+Note that in this alternative we can only have one test for a test and query files.
+
+## Design
+
+This is an independent command line tool for testing KSQL statement by evaluating the generated results for a given input data. Test settings along with the input data and expected output are passed to the tool through a test file in JSON format.
```

Comment:
We should also support an 'embedded mode', where users can use the library code directly in a Java build pipeline to run their tests if they so want.

This basically comes for free. The only cost associated is that we need to ensure API compatibility on the specific class(es) they be using as we enhance from release to release.